core/repository/sparql/src/main/java/org/eclipse/rdf4j/repository/sparql/federation/RepositoryFederatedService.java

```diff
@@ -87,7 +115,24 @@ protected void handleBindings() throws Exception {
 				}
 				CloseableIteration<BindingSet> materializedIter = new CollectionIteration<>(
 						blockBindings);
-				addResult(evaluateInternal(service, materializedIter, service.getBaseURI()));
+				// The request queue is of unbounded size, so this will not block. On the other hand,
+				// if the consumer is slow, we may accumulate a large number of pending requests here.
+				// I don't have a better solution at the moment, though.
+				requestQueue.put(materializedIter);
```

Comment:
I have a question about spilling to disk. It's probably out of scope for this PR, since you are primarily fixing a bug. 

Would this be a good place in the code to check for low mem and spill to disk?

Reply:
Yes, it would, I think. If we had a huge intermediate result set, we could store the queue on disk.